### PR TITLE
chore: remove knative auto controller enablement

### PR DESCRIPTION
**Pull request by shaneutt**
This routine was previously started as we had an understanding
of the V1 implementation that it was automatically and dynamically
enabling knative support on the fly. However after re-reading the
V1 manager code, it turns out this is not the case and so having
this functionality is effectively a new feature. This patch removes
this feature so that we don't add it to the scope and maintenance
of the upcoming 2.0 release as there weren't any requirements for
it at the time, and to keep maintenance and scope tight.

We will re-asses proactive and dynamically loaded controllers in
a later iteration, see:

https://github.com/Kong/kubernetes-ingress-controller/issues/1449

### CODE REVIEW

**Comment by shaneutt** [comment]
@ccfishk is a critical stakeholder in this and must sign off.

**Comment by ccfishk** [comment]
@shaneutt  Knative support is K2.0 parity of v1.3.x - support Knative CRD dynamically. I don't see any reason of removing this function.

**Comment by shaneutt** [comment]
> @shaneutt Knative support is K2.0 parity of v1.3.x - support Knative CRD dynamically. I don't see any reason of removing this function.

In `v1.3.x` we _do not dynamically enable knative support if the CRD becomes available_. I think the problem is that the old manager code is very terse, and it kinda looks like it _does_ but @mflendrich and I tracked through the code and it in fact does not.

**Comment by ccfishk** [comment]
> > @shaneutt Knative support is K2.0 parity of v1.3.x - support Knative CRD dynamically. I don't see any reason of removing this function.
> 
> In `v1.3.x` we _do not dynamically enable knative support if the CRD becomes available_. I think the problem is that the old manager code is very terse, and it kinda looks like it _does_ but @mflendrich and I tracked through the code and it in fact does not.

Well, this dynamic enablement at least erase following two works
- we dont maintain third party CRD in kong repo
- dont need to restart KIC to support CRD

As you are already aware, drop off adding CRDs (i.e plugin, consumer), migration to add annotations in ingress directly would be our final goal in GateWayAPI project.
I dont see the necessity to remove this feature.

**Comment by rainest** [comment]
As I understand it, there are two options to ensure that controllers don't run if missing CRDs means we should disable them automatically:

- Override the flag value when adding the controller to the controller list. Set [`IsEnabled`](https://github.com/Kong/kubernetes-ingress-controller/blob/cb27c4b2d83abb068c55d03a16f0602a77f89cae/railgun/manager/run.go#L190) to a value other than what's in the config struct. This is (roughly) the approach used by the code this PR will remove.
- Do not add the controller to the controller list at all. Check a condition ("is CRD present?") and [append the controller to the controller list if that condition is satisfied or do nothing if the condition is not satisfied](https://github.com/Kong/kubernetes-ingress-controller/blob/cb27c4b2d83abb068c55d03a16f0602a77f89cae/railgun/manager/run.go#L340-L357). This approach was used in 1.x and was [what I used when adding functionality to disable KongClusterPlugins](https://github.com/Kong/kubernetes-ingress-controller/commit/4f962953e39ba3eab934234b63ac259388ad16d1). I also added similar logic to add or not add the Knative controller.

I think either approach is valid and don't think there's any practical difference between loading a controller (not including it in the [controller list](https://github.com/Kong/kubernetes-ingress-controller/blob/cb27c4b2d83abb068c55d03a16f0602a77f89cae/railgun/manager/run.go#L359-L363)), but marking it disabled, and not loading a controller. I went with the latter approach because I was familiar with it from 1.x. There could be some subtle difference, but either approach means the controller doesn't run if its CRDs are missing.

Using both approaches is redundant, so the change proposed here just removes one of the approaches because you don't need both--the controller doesn't run either way--and pruning it simplifies the codebase.

We may want to refactor how this works if we end up needing to auto-disable more controllers (we could make the code more generic), but for now that's YAGNI.

**Comment by ccfishk** [comment]
Generically, a controller/operator, run with predefined scope - supported CRD, if exists, setup controller with manager, run controllers. 

Or, supported CRDs; if some CRDs not exist, still run, bump controllers on the fly. A generic function/utility implementing this would be nice. 
Assuming we want latter.

**Comment by ccfishk** [comment]
Offline discussion - Keep Knative feature as it is in 2.0. Two fixes are along:
Fix the disablement setting case https://github.com/Kong/kubernetes-ingress-controller/pull/1550
Nonfunction lint errors https://github.com/Kong/kubernetes-ingress-controller/pull/1537

**Comment by shaneutt** [comment]
> Offline discussion - Keep Knative feature as it is in 2.0. Two fixes are along:
> Fix the disablement setting case #1550
> Nonfunction lint errors #1537

Wait, sorry, there's been some kind of a miscommunication perhaps in some nuance of things.

I want to be very clear on my position:

1. I think that there's a lot of _value_ in an autoloading controller feature for our end users
2. I'm still not certain what I think is the best _timing_ for delivering that: it seems super safe to bring it back in post-2.0, but it seems like we keep growing out 2.0 scope by trying to include it.

Again the value is not in question _at all_ I really like the feature idea, I do however think as a team we on @Kong/team-k8s need to come to a consensus on the _when_ part, which I can not answer by myself.

**Review by rainest — APPROVED** [review]
To affirm my position here, approving since I'm fine with this change and don't think I have anything further to add re why it's okay (we had two systems for managing controller auto-loading, this prunes one of them without changing anything re 1.x behavior, and we can revisit additional refactoring or behavior changes post-2.x).

If I can provide any additional historical context or review that would help for consensus, please @ me.

File: railgun/manager/run.go [hunk 1]
```diff
@@ -382,49 +375,6 @@ func Run(ctx context.Context, c *config.Config) error {
 		setupLog.Info("anonymous reports disabled, skipping")
 	}
 
-	go FlipKnativeController(mgr, prx, &c.KnativeIngressEnabled, c, setupLog)
 	setupLog.Info("starting manager")
 	return mgr.Start(ctx)
 }
-
-// wait for knative cr before register and starting knative controller
-func FlipKnativeController(mgr manager.Manager, prx proxy.Proxy, enablestatus *util.EnablementStatus, cfg *config.Config, log logr.Logger) error {
```
Comment (ccfishk) [inline railgun/manager/run.go:391]: We should add the lint error as we discussed instead of remove the function

**Review by ccfishk — COMMENTED** [review]
left a comment

**Comment by ccfishk** [comment]
We do not even have a syncup within the team as planned/discussed on slack.
Please do not merge this PR.

**Comment by shaneutt** [comment]
> We do not even have a syncup within the team as planned/discussed on slack.
> Please do not merge this PR.

Don't worry, as I said at the beginning:

https://github.com/Kong/kubernetes-ingress-controller/pull/1524#issuecomment-877451694

I consider you a _critical stakeholder_ in this having originally worked on the feature, and I have _no intention to merge without your approval_ :bow:

**Comment by mflendrich** [comment]
The decision made by the team is that we do not include dynamic loading of knative controller in 2.0.

**Comment by shaneutt** [comment]
Given long time this PR has been sitting and the conflicts with next, I opted for a fresh commit/PR: https://github.com/Kong/kubernetes-ingress-controller/pull/1571

**Pull request closed without merging**